### PR TITLE
feat(authentication-service): maintain the entries for active users when a user login to the app

### LINKED ISSUE #1424: Add ability to count daily active users
**Is your feature request related to a problem? Please describe.**
Currently there is no provision to a count of daily active users of the application.

**Describe the solution you'd like**
Save the daily active users
Whenever a user logins in to system mark the user as active with the timestamp

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by yeshamavani**
gh-1424

BREAKING CHANGE
Added migrations

## Description

1) When a user log in or logout of the application an entry is made in to table login_activity.
2) We mark three activities based on user action - ACCESS, RELOGIN, LOGOUT.
3) while me mark the user activity we save other information about the user alon with its activity type.
3) this marking of activity is configurable via the binding key --> AuthServiceBindings.MarkUserActivity that takes a boolean value .
4) Once we have logged the users activity based on application's need we can get Daily or Monthly active uses. This gives as a list of unique active users for the provided start and end date for the selected type -- daily or monthly.

Fixes #1424 

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Intermediate change (work in progress)

## Checklist:

- [x] Performed a self-review of my own code
- [ ] npm test passes on your machine
- [ ] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the style guide
- [ ] API Documentation in code was updated
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: services/authentication-service/migrations/sqls/20230523175343-active-users-up.sql [hunk 1]
```diff
@@ -0,0 +1,10 @@
+/* Replace with your SQL commands */
+
+CREATE TABLE main.active_users (
```
Comment (samarpanB) [inline services/authentication-service/migrations/sqls/20230523175343-active-users-up.sql:3]: table name is wrong. It should be named as login_activity

File: services/authentication-service/migrations/sqls/20230523175343-active-users-up.sql [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+		id 						uuid DEFAULT md5(random()::text || clock_timestamp()::text)::uuid NOT NULL ,
+		actor                   text NOT NULL,
+		tenant_id				text NOT NULL,
```
Comment (samarpanB) [inline services/authentication-service/migrations/sqls/20230523175343-active-users-up.sql:6]: tenant id may not be there for every app

File: services/authentication-service/migrations/sqls/20230523175343-active-users-up.sql [hunk 3]
```diff
@@ -0,0 +1,10 @@
...
+		login_time				timestamptz DEFAULT CURRENT_TIMESTAMP NOT NULL,
```
Comment (samarpanB) [inline services/authentication-service/migrations/sqls/20230523175343-active-users-up.sql:7]: We need to track more data like, browser, ip address (encrypted), success or error, etc. Refer this - https://www.databasezone.com/samples/User/index.htm#LoginAttempt

File: services/authentication-service/src/controllers/active-users.controller.ts [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+  CONTENT_TYPE,
+  OPERATION_SECURITY_SPEC,
+  STATUS_CODE,
+} from '@sourceloop/core';
+import {ActiveUsers} from '../models';
+import {PermissionKey} from '../enums';
+
+const baseUrl = '/active-users';
+
+export class ActiveUsersController {
```
Comment (samarpanB) [inline services/authentication-service/src/controllers/active-users.controller.ts:28]: This entire controller does not make sense at all. It cannot be active users. For active users u need to add BL as well.

File: services/authentication-service/src/component.ts [hunk 5]
```diff
@@ -163,6 +163,7 @@ export class AuthenticationServiceComponent implements Component {
       .bind('services.LoginHelperService')
       .toClass(LoginHelperService);
     this.application.bind('services.otpService').toClass(OtpService);
+    this.application.bind(AuthServiceBindings.ActorIdKey).to('id');
```
Comment (samarpanB) [inline services/authentication-service/src/component.ts:166]: why 'id' ?

File: services/authentication-service/src/models/active-users.model.ts [hunk 6]
```diff
@@ -0,0 +1,43 @@
+import {Entity, model, property} from '@loopback/repository';
+
+@model({
+  description: `This is to maintain the daily active users list.`,
+  settings: {strict: true},
```
Comment (samarpanB) [inline services/authentication-service/src/models/active-users.model.ts:5]: remove this

File: services/authentication-service/src/modules/auth/login.controller.ts [hunk 7]
```diff
@@ -490,6 +495,22 @@ export class LoginController {
         },
         {ttl: authClient.refreshTokenExpiration * ms},
       );
+
+      // make an entry to mark the active user
+
+      const actor = user[this.actorKey]?.toString() ?? '0';
```
Comment (samarpanB) [inline services/authentication-service/src/modules/auth/login.controller.ts:501]: We need to track access and refresh event separately

**Review by samarpanB — CHANGES_REQUESTED** [review]

File: services/authentication-service/migrations/sqls/20230523175343-active-users-up.sql (hunk 2 above)
Comment (yeshamavani) [reply services/authentication-service/migrations/sqls/20230523175343-active-users-up.sql:6]: then should we add the defaultTenantId?

File: services/authentication-service/src/controllers/active-users.controller.ts (hunk 4 above)
Comment (yeshamavani) [reply services/authentication-service/src/controllers/active-users.controller.ts]: done

File: services/authentication-service/src/component.ts (hunk 5 above)
Comment (yeshamavani) [reply services/authentication-service/src/component.ts]: default will be userId but the field is configurable.

File: services/authentication-service/src/models/active-users.model.ts (hunk 6 above)
Comment (yeshamavani) [reply services/authentication-service/src/models/active-users.model.ts]: done

File: services/authentication-service/src/modules/auth/login.controller.ts (hunk 7 above)
Comment (yeshamavani) [reply services/authentication-service/src/modules/auth/login.controller.ts]: tracking access, relogin and logout separately

File: services/authentication-service/src/controllers/login-activity.controller.ts [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+        content: {
+          [CONTENT_TYPE.JSON]: {
+            schema: getModelSchemaRef(Object, {includeRelations: true}),
+          },
+        },
+      },
+    },
+  })
+  async getActiveUsers(
+    @param.path.string('range') range: string,
```
Comment (samarpan-b) [inline services/authentication-service/src/controllers/login-activity.controller.ts:137]: use enum for range

File: services/authentication-service/src/controllers/login-activity.controller.ts [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+      },
+    });
+
+    //all the BL here
+
+    const groupByDate: ActiveUsersGroupData = {};
+    activeUsersForTime.forEach(item => {
+      let date = '';
+      if (range === 'daily') {
+        date = item.loginTime.toISOString().split('T')[0];
```
Comment (samarpan-b) [inline services/authentication-service/src/controllers/login-activity.controller.ts:154]: use moment for date-time calculations

**Review by samarpan-b — CHANGES_REQUESTED** [review]

File: services/authentication-service/src/controllers/login-activity.controller.ts (hunk 8 above)
Comment (yeshamavani) [reply services/authentication-service/src/controllers/login-activity.controller.ts:137]: done

File: services/authentication-service/src/modules/auth/logout.controller.ts [hunk 10]
```diff
@@ -110,6 +136,35 @@ export class LogoutController {
...
+      loginTime: new Date(),
+      tokenPayload: {...user, clientId: refreshTokenModel.clientId},
+      deviceInfo: this.ctx.request.headers['user-agent']?.toString(),
+      loginType: LoginType.LOGOUT,
+      ipAddress:
+        this.ctx.request.headers['x-forwarded-for']?.toString() ??
+        this.ctx.request.socket.remoteAddress,
+    });
+    this.loginActivityRepo.create(loginActivity).catch(e => {
+      console.log(e.toString());
```
[the remark this answers is not in the document]
Comment (yeshamavani) [reply services/authentication-service/src/modules/auth/logout.controller.ts:163]: not needed was for testing 
removed.

File: services/authentication-service/src/models/login-activity.model.ts [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+    default: () => new Date(),
+    name: 'login_time',
+  })
+  loginTime: Date;
+
+  @property({
+    type: 'object',
+    name: 'token_payload',
+  })
+  tokenPayload?: object;
```
[the remark this answers is not in the document]
Comment (yeshamavani) [reply services/authentication-service/src/models/login-activity.model.ts]: done

File: services/authentication-service/src/models/login-activity.model.ts [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+    type: 'string',
+    name: 'device_info',
+  })
+  deviceInfo: string;
+
+  @property({
+    type: 'string',
+    name: 'ip_address',
+  })
+  ipAddress: string;
```
[the remark this answers is not in the document]
Comment (yeshamavani) [reply services/authentication-service/src/models/login-activity.model.ts:56]: done

File: services/authentication-service/src/controllers/login-activity.controller.ts (hunk 9 above)
Comment (yeshamavani) [reply services/authentication-service/src/controllers/login-activity.controller.ts:154]: done

File: services/authentication-service/src/controllers/login-activity.controller.ts [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+    const groupByDate: ActiveUsersGroupData = {};
+    activeUsersForTime.forEach(item => {
+      let date = '';
+      const loginTime = moment(item.loginTime);
+      if (range === ActiceUsersRange.DAILY) {
+        date = loginTime.format('YYYY-MM-DD');
+      } else if (range === ActiceUsersRange.MONTHLY) {
+        const month = loginTime.month() + 1;
+        const year = loginTime.year();
+        date = `${year}-${month}`;
```
Comment (samarpan-b) [inline services/authentication-service/src/controllers/login-activity.controller.ts:160]: You can use format here too

File: services/authentication-service/src/enums/login-type.enum.ts [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+// This software is released under the MIT License.
+// https://opensource.org/licenses/MIT
+
+export const enum LoginType {
+  ACCESS = 'ACCESS',
+  RELOGIN = 'RELOGIN',
+  LOGOUT = 'LOGOUT',
+}
+
+export const enum ActiceUsersRange {
```
Comment (samarpan-b) [inline services/authentication-service/src/enums/login-type.enum.ts:12]: Spell check please

File: services/authentication-service/src/modules/auth/login.controller.ts [hunk 15]
```diff
@@ -472,11 +488,11 @@ export class LoginController {
           AuthenticateErrorKeys.UserDoesNotExist,
         );
       }
-      const data = await this.getJwtPayload(user, authClient);
+      const data: AnyObject = await this.getJwtPayload(user, authClient);
```
Comment (samarpan-b) [inline services/authentication-service/src/modules/auth/login.controller.ts:491]: Why type added here ?

File: services/authentication-service/src/modules/auth/login.controller.ts [hunk 16]
```diff
@@ -490,6 +506,13 @@ export class LoginController {
         },
         {ttl: authClient.refreshTokenExpiration * ms},
       );
+
+      const userTenant = await this.userTenantRepo.findOne({
+        where: {userId: user.id},
+      });
+
+      this.markUserActivity(user, userTenant, {...data}, loginType);
```
Comment (samarpan-b) [inline services/authentication-service/src/modules/auth/login.controller.ts:514]: Lets make this feature configurable. Take a boolean in component config and use that to check here

File: services/authentication-service/src/modules/auth/login.controller.ts [hunk 17]
```diff
@@ -507,4 +530,64 @@ export class LoginController {
...
+  private markUserActivity(
+    user: User,
+    userTenant: UserTenant | null,
+    payload: AnyObject,
+    loginType: LoginType,
+  ) {
+    const size = 16;
+    const encryptionKey = process.env.ENCRYPTION_KEY;
+
+    if (encryptionKey) {
```
Comment (samarpan-b) [inline services/authentication-service/src/modules/auth/login.controller.ts:543]: In next PR, move this encryption part to a provider and let consumer override it if needed.

File: sandbox/auth-ms-basic-example/package.json [hunk 18]
```diff
@@ -70,6 +70,7 @@
     "db-migrate-pg": "^1.2.2",
     "dotenv": "^16.0.3",
     "dotenv-extended": "^2.9.0",
+    "i": "^0.3.7",
```
[the remark this answers is not in the document]
Comment (yeshamavani) [reply sandbox/auth-ms-basic-example/package.json:73]: mistake

File: services/authentication-service/src/controllers/login-activity.controller.ts (hunk 13 above)
Comment (yeshamavani) [reply services/authentication-service/src/controllers/login-activity.controller.ts:160]: done

File: services/authentication-service/src/enums/login-type.enum.ts (hunk 14 above)
Comment (yeshamavani) [reply services/authentication-service/src/enums/login-type.enum.ts:12]: done

File: sandbox/auth-ms-basic-example/package.json (hunk 18 above)
[the remark this answers is not in the document]
Comment (samarpan-b) [reply sandbox/auth-ms-basic-example/package.json:73]: its still here

**Review by samarpan-b — APPROVED** [review]

**Pull request merged**